src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -321,20 +327,100 @@ isAcyclic = isRight . topSort
 -- 'isAcyclic' x     == (scc x == 'gmap' NonEmpty.'NonEmpty.vertex' x)
 -- @
 scc :: Ord a => AdjacencyMap a -> AdjacencyMap (NonEmpty.AdjacencyMap a)
-scc m = gmap (component Map.!) $ removeSelfLoops $ gmap (leader Map.!) m
+scc g = condense g $ execState (gabowSCC g) initialState where
+  initialState = C 0 0 [] [] Map.empty Map.empty IntMap.empty [] []
+
+data StateSCC a
+  = C { current       :: !Int
+      , componentId   :: !Int
+      , boundary      :: ![(Int,a)]
+      , dfsPath       :: ![a]
+      , preorders     :: !(Map.Map a Int)
+      , components    :: !(Map.Map a Int)
+      , inner_graphs  :: (IntMap.IntMap (List (AdjacencyMap a)))
+      , inner_edges   :: ![(Int,(a,a))]
```

Comment:
Are you sure all the tuples here (and in other fields) have the right strictness? While the list itself is marked as strict, the inner tuples may turn out to be lazy. Is this what you want, or are you sure GHC correctly optimises laziness away?

Reply:
I'm not sure at all if strictness matters in this case, but in the `Int` version we would probably want to at least unpack all these tuples, right? Otherwise, we may be constantly chasing pointers.

Reply:
I'm not sure if the amount of laziness makes sense. I experimented with using `StrictPair` and saw a fairly big regression in the benchmarks. I then tried changing the fields (except counters) to lazy and saw an improvement across the board!

Reply:
My original reasoning was along the lines of if a whole structure is eventually consumed, it is faster when strict

Reply:
Interesting! I guess one example where "if a whole structure is eventually consumed, it is faster when strict" doesn't work is when intermediate lists are eliminated thanks to list fusion. Perhaps, this is what happened here.